### PR TITLE
feat: add custom replacer function

### PR DESCRIPTION
**Pull request by semmel**
- perhaps the naming could be improved
- I didn't feel like supporting array too as replacer — as in `JSON.stringify`

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -7,6 +7,12 @@ export type Options = {
 	Set it to `'\t'` for tab indentation or the number of spaces you want.
 	*/
 	readonly indentation?: string | number;
+	/**
+	 Replacer function applied before circularity is detected.
+
+	 see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
+	 */
+	customReplacer?: (key: string | number, value: any) => any;
```
Comment (sindresorhus) [inline index.d.ts]: Why not?
```suggestion
	replacer?: (key: string | number, value: any) => any;
```

File: index.d.ts [hunk 2]
```diff
@@ -7,6 +7,12 @@ export type Options = {
 	Set it to `'\t'` for tab indentation or the number of spaces you want.
 	*/
 	readonly indentation?: string | number;
+	/**
+	 Replacer function applied before circularity is detected.
```
Comment (sindresorhus) [inline index.d.ts]: This should explain a little bit more what a replacer function does.

File: index.d.ts [hunk 3]
```diff
@@ -7,6 +7,12 @@ export type Options = {
...
+
+	 see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
```
Comment (sindresorhus) [inline index.d.ts]: Use the direct link: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#replacer

File: index.d.ts [hunk 4]
```diff
@@ -7,6 +7,12 @@ export type Options = {
...
+	 */
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	*/
```

File: index.d.ts (hunk 1 above)
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	customReplacer?: (key: string | number, value: unknown) => any;
```

File: readme.md [hunk 5]
```diff
@@ -23,6 +23,13 @@ console.log(safeStringify(foo));
 
 console.log(JSON.stringify(foo));
 //=> TypeError: Converting circular structure to JSON
+
+//Use customReplacer option to support BigInt
+const bar = {x: 123n};
+bar.y = bar;
+
+console.log(safeStringify(foo, {customReplacer: (k, v) => typeof v === 'bigint' ? `${v.toString()}n` : v}));
+//=> '{"x":"123n","y":"[Circular]"}'
```
Comment (sindresorhus) [inline readme.md]: Move this code example to line 66.

File: readme.md [hunk 6]
```diff
@@ -23,6 +23,13 @@ console.log(safeStringify(foo));
 
 console.log(JSON.stringify(foo));
 //=> TypeError: Converting circular structure to JSON
+
+//Use customReplacer option to support BigInt
+const bar = {x: 123n};
+bar.y = bar;
+
+console.log(safeStringify(foo, {customReplacer: (k, v) => typeof v === 'bigint' ? `${v.toString()}n` : v}));
```
Comment (sindresorhus) [inline readme.md]: Don't use one character parameter names.

File: readme.md [hunk 7]
```diff
@@ -23,6 +23,13 @@ console.log(safeStringify(foo));
 
 console.log(JSON.stringify(foo));
 //=> TypeError: Converting circular structure to JSON
+
+//Use customReplacer option to support BigInt
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
// Use `customReplacer` option to support BigInt.
```

File: readme.md [hunk 8]
```diff
@@ -51,6 +56,14 @@ The indentation of the JSON.
 
 By default, the JSON is not indented. Set it to `'\t'` for tab indentation or the number of spaces you want.
 
+##### customReplacer
+
+Type: `(key: string, value: any) => any`
+
+A replacer as for [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) to alter the serialization for custom objects.
+
+The custom replacer is applied before circularity is detected.
```
Comment (sindresorhus) [inline readme.md:58]: The docs here should be in sync with the one in index.d.ts

File: test.js [hunk 9]
```diff
@@ -216,3 +216,50 @@ test('array containing objects with the same circular reference', t => {
 
 	t.snapshot(safeStringify(fixture, options));
 });
+
+test('circular object w BigInt custom replacer', t => {
```
Comment (sindresorhus) [inline test.js]:
```suggestion
test('circular object with BigInt custom replacer', t => {
```

File: index.js [hunk 10]
```diff
@@ -22,7 +22,16 @@ function safeStringifyReplacer(seen) {
...
 	const seen = new WeakSet();
-	return JSON.stringify(object, safeStringifyReplacer(seen), indentation);
+	let preReplacer = (k, v) => v;
+	if (customReplacer) {
+		preReplacer = customReplacer;
+	}
+
+	return JSON.stringify(
+		object,
+		(key, value) => safeStringifyReplacer(seen)(key, preReplacer.call(object, key, value)),
```
Comment (sindresorhus) [inline index.js:34]: Wouldn't it be simpler to pass the custom replaces into `safeStringifyReplacer` and call it there?

File: index.js [hunk 11]
```diff
@@ -22,7 +22,16 @@ function safeStringifyReplacer(seen) {
 	};
 }
 
-export default function safeStringify(object, {indentation} = {}) {
+export default function safeStringify(object, {indentation, customReplacer} = {}) {
 	const seen = new WeakSet();
-	return JSON.stringify(object, safeStringifyReplacer(seen), indentation);
+	let preReplacer = (k, v) => v;
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	let preReplacer = (key, value) => value;
```

**Comment by sindresorhus** [comment]
> I didn't feel like supporting array too as replacer — as in JSON.stringify

That needs to be documented.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js (hunk 10 above)
Comment (semmel) [reply index.js:34]: Perhaps, but `safeStringifyReplacer` is a function that works correctly. I didn't want to change its signature nor implementation, not knowing were if it is used anywhere else — it is not, I know.

**Comment by semmel** [comment]
Hope, I have addressed all your points with https://github.com/semmel/safe-stringify/commit/4d7bdad6c36c923d1ececfedf821560b0256dbf3

File: readme.md [hunk 12]
```diff
@@ -51,6 +49,28 @@ The indentation of the JSON.
...
+```js
+const bar = {x: 123n};
+bar.y = bar;
+
+console.log(safeStringify(
+	foo,
+	{replacer: (key, value) => typeof value === 'bigint' ? `${value.toString()}n` : value}
+));
+//=> '{"x":"123n","y":"[Circular]"}'
+```
```
Comment (sindresorhus) [inline readme.md:72]: This is missing from index.d.ts. Use `@example` tag.

File: readme.md [hunk 13]
```diff
@@ -51,6 +49,28 @@ The indentation of the JSON.
...
+
+A replacer function as for [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#replacer) to alter the serialization for custom objects.
+
+The custom replacer is applied before circularity is detected.
+
+*Note: Unlike `JSON.stringify` `replacer` cannot be an array.*
+
+Example: Use `replacer` option to support BigInt.
+```js
+const bar = {x: 123n};
```
Comment (sindresorhus) [inline readme.md:64]: include import statement so the code example is runnable.

File: readme.md [hunk 14]
```diff
@@ -51,6 +49,28 @@ The indentation of the JSON.
 
 By default, the JSON is not indented. Set it to `'\t'` for tab indentation or the number of spaces you want.
 
+##### replacer
+
+Type: `(key: string, value: unknow) => any`
+
+A replacer function as for [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#replacer) to alter the serialization for custom objects.
```
Comment (sindresorhus) [inline readme.md:56]: This sentence does not read well.

Tip: Run it through ChatGPT.

File: index.js [hunk 15]
```diff
@@ -22,7 +22,16 @@ function safeStringifyReplacer(seen) {
 	};
 }
 
-export default function safeStringify(object, {indentation} = {}) {
+export default function safeStringify(object, {indentation, replacer} = {}) {
```
Comment (sindresorhus) [inline index.js:25]:
```suggestion
export default function safeStringify(object, {indentation, replacer = (key, value) => value} = {}) {
```

File: index.js (hunk 10 above)
Comment (sindresorhus) [reply index.js:34]: Ok

File: readme.md [hunk 16]
```diff
@@ -51,6 +49,28 @@ The indentation of the JSON.
...
+
+The custom replacer is applied before circularity is detected.
+
+*Note: Unlike `JSON.stringify` `replacer` cannot be an array.*
```
Comment (sindresorhus) [inline readme.md:60]:
```suggestion
*Note: Unlike `JSON.stringify`, the `replacer` option here cannot be an array.*
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts [hunk 17]
```diff
@@ -7,6 +7,14 @@ export type Options = {
 	Set it to `'\t'` for tab indentation or the number of spaces you want.
 	*/
 	readonly indentation?: string | number;
+	/**
```
Comment (sindresorhus) [inline index.d.ts:10]:
```suggestion

	/**
```

**Comment by sindresorhus** [comment]
Bump :)

**Comment by semmel** [comment]
> Bump :)

though remaining changes seem minor, there is few time here. The weekend in a fortnight  should do.

**Comment by sindresorhus** [comment]
Bump

**Comment by semmel** [comment]
> Bump

sorry, I am afraid that I won't have enough time to complete th PR.
It was a problem — I thought — could be resolved swiftly in passing. But it takes more time than expected.
Thanks for your response anyway.
With regards Semmel

**Pull request closed without merging**